Fixes to support newer pinecone sdk

manushreegangwar opened the pull request:
This PR is to add support for newer pinecone SDK: https://docs.pinecone.io/reference/python-sdk

manushreegangwar commented on tests/intensive/test_similarity.py:
```diff
@@ -235,13 +237,13 @@ def test_image_similarity_backends():
 
-        index2 = dataset.load_brain_results(brain_key)
-        assert index2.total_index_size == 200
+        # index2 = dataset.load_brain_results(brain_key)
```
@brimoor Do you know understand `index2._index = None` when using `load_brain_results`?

manushreegangwar replied:
This happens when we call `dataset.clean_cache` before `load_brain_results`. 

I have changed the order as a workaround. Let me know if you have a cleaner solution.

brimoor replied:
There shouldn't need to be any reordering of commands in the tests; that indicates that the updated PR doesn't quite behave the same as the previous implementation did.

The intent of `dataset.load_brain_results(brain_key)` is that it establishes a connection to the existing Pinecone index if one was previously created. Take a look at the implementation and see if you can spot why the behavior changed from the previous implementation where the tests were passing.

brimoor replied:
To clarify: if you call `dataset.load_brain_results(brain_key)` in the same Python session where you originally called `compute_similarity(dataset, brain_key=brain_key, ...)`, it will return the cached object, which has already taken care of initializing `_index` (indeed, it created the index).

However, each `SimilarityIndex` class also needs to support the workflow where a user wants to connect to an existing similarity index on an existing dataset in a fresh Python session:

```py
import fiftyone as fo

dataset = fo.load_dataset(...)
index = dataset.load_brain_results(brain_key)

assert index._index is not None
```

manushreegangwar replied:
Yes, `dataset.clean_cache` destroys `index._index` for pinecone and `load_brain_results` doesn't load it. There's no `index._index` in sklearn backend which is why the test passes for sklearn.

I believe this behavior precedes the PR. I will dig around to see how to fix the issue.

manushreegangwar commented on fiftyone/brain/internal/core/pinecone.py:
```diff
@@ -31,6 +32,8 @@ class PineconeSimilarityConfig(SimilarityConfig):
 
     Args:
+        cloud ("aws"): Cloud where Pinecone index would be hosted
+        environment ("us-east-1"): a Pinecone environment to use
```
https://docs.pinecone.io/reference/api/2024-07/control-plane/create_index

These are required parameters and pinecone doesn't have any defaults. I am setting these as the default values which are overwritten at config initialization so I wouldn't consider this hard-coding. I can change the implementation to handle this in `create_index`.

I will move them to the bottom of the list.

manushreegangwar commented on fiftyone/brain/internal/core/pinecone.py:
```diff
@@ -183,17 +191,29 @@ class PineconeSimilarityIndex(SimilarityIndex):
-            index_names = pinecone.list_indexes()
+            index_names = (
+                self._pinecone.list_indexes()
+                if self._pinecone
```
This will cause issues in `create_index`

manushreegangwar commented on fiftyone/brain/internal/core/pinecone.py:
```diff
@@ -209,30 +229,76 @@ def _initialize(self):
+            while not self._pinecone.describe_index(
+                self.config.index_name
+            ).status["ready"]:
+                time.sleep(1)
```
Yes, `pinecone` takes times to initialize and I was running into issues because of that. Agree with the no timeout in core but I didn't find a method in pinecone SDK for this. I will check again to see if there's an alternative

brimoor replied:
The MongoDB backend has a similar behavior so we added a `ready` property:
https://github.com/voxel51/fiftyone-brain/blob/2374ef5b20985ad65b38665d55effd1cce725a7a/fiftyone/brain/internal/core/mongodb.py#L282-L301

That can be used as documented here:
https://docs.voxel51.com/integrations/mongodb.html#checking-if-an-index-is-ready

Can you adopt that pattern here for consistency?

manushreegangwar commented on tests/intensive/test_similarity.py:
```diff
@@ -164,6 +164,8 @@ def test_image_similarity_backends():
 
     # sklearn backend
     ###########################################################################
+    if "clip_sklearn" in dataset.list_brain_runs():
```
These changes were made to handle the tests not reaching `dataset.delete()`. I personally don't like implicit assumption when running tests. I didn't know about `drop_existing_dataset=True`. Thanks for the suggestion. I will try it out.

manushreegangwar commented:
Thanks for the feedback, @brimoor. I have addressed your comments and  all "pinecone" related tests in test_similarity.py are passing. Ready for review :mag:

brimoor commented on fiftyone/brain/internal/core/pinecone.py:
```diff
@@ -183,17 +191,29 @@ class PineconeSimilarityIndex(SimilarityIndex):
-            environment=self.config.environment,
-            project_name=self.config.project_name,
-        )
+        try:
```
I believe that `pinecone.init()` was removed in pinecone v3. That's multiple versions ago, so after a bit more thought, I think it would be best to just require:

```py
    def ensure_usage_requirements(self):
        fou.ensure_package("pinecone-client>=3")
```

so that we can remove the `if` statements in this implementation and keep it clean. Sound good to you?

brimoor replied:
Please double check that `pinecone-client>=3` is in fact correct 😄

manushreegangwar replied:
Done. Anything >= 3.2 should be fine. 

`pinecone-client==3.0` is a mystery as it neither has `init` not `Pinecone` :shrug:

brimoor commented on fiftyone/brain/internal/core/pinecone.py:
```diff
@@ -243,9 +309,29 @@ def index(self):
+            ).status["ready"]
+        return pinecone.describe_index(self.config.index_name).status["ready"]
+
+    def verify_total_index_size(self, expected_size, timeout=10, interval=1):
```
It's unfortunate that Pinecone itself doesn't have a concept of what MongoDB calls "write concern" that allows you to wait for their product to reflect the changes that were requested 😢

In practice, when potentially multiple people are working on the same Pinecone index, it feels a bit hacky to wait for the index to reach a certain size. Maybe someone else deleted a sample from the index while this was polling, for example. The ideal thing would be "wait until Pinecone acknowledges that my current transaction completes". But that's not something we should build if Pinecone doesn't support it.

In my opinion, this feels more like a util function that we should put in the `tests` folder to start with so that our integration tests pass (it could be used by any backend, FWIW).

brimoor replied:
If you feel strongly that this is a util we should expose to users, then we can keep it here. Up to you 😄

manushreegangwar replied:
:+1: Moved this to `test_similarity.py`.

brimoor reviewed (changes requested):
Almost there! Great updates so far 💪

manushreegangwar commented on fiftyone/brain/internal/core/pinecone.py:
```diff
@@ -18,6 +18,7 @@
     SimilarityIndex,
 )
 import fiftyone.brain.internal.core.utils as fbu
+from pinecone import Pinecone, ServerlessSpec, PodSpec
```
I didn't notice `lazy_import`. Thanks, switched pinecone to use that.

brimoor reviewed (approved):
LGTM
